tests/functional/validation.py

```diff
@@ -0,0 +1,95 @@
+import concurrent.futures
+import csv
+import io
+import os
+import shutil
+import tempfile
+
+import hca.dss
+import loompy
+import numpy
+import requests
+import s3fs
+import zarr
+
+
+DSS_CLIENT = hca.dss.DSSClient()
+DSS_CLIENT.host = f"https://dss.{os.environ['DEPLOYMENT_STAGE']}.data.humancellatlas.org/v1"
+
+S3 = s3fs.S3FileSystem(anon=True)
+
+
+def calculate_ss2_metrics_direct(bundle_fqids):
+    """Calculate expected SS2 matrix values.
+
+    Don't use matrices or the matrix service, calculate in a completely orthogonal way
+    using the RSEM outputs directly.
+    """
+
+    def read_bundle(fqid):
+        bundle_uuid, bundle_version = fqid.split(".", 1)
+        bundle = DSS_CLIENT.get_bundle(uuid=bundle_uuid, version=bundle_version, replica="aws")
+
+        rsem_file = [f for f in bundle['bundle']['files']
+                     if f["name"].endswith(".genes.results")][0]
+
+        rsem_contents = DSS_CLIENT.get_file(uuid=rsem_file["uuid"],
+                                            version=rsem_file["version"],
+                                            replica="aws")
+
+        rsem_reader = csv.DictReader(io.StringIO(rsem_contents.decode()), delimiter="\t")
+
+        bundle_expression_sum = 0
+        bundle_expression_nonzero = 0
+        for row in rsem_reader:
+            bundle_expression_sum += float(row["TPM"])
+            if float(row['TPM']) != 0.0:
+                bundle_expression_nonzero += 1
+        return {"expression_sum": bundle_expression_sum,
```

Comment:
nit: single quotes for keys

Reply:
Are you saying you want more single quotes or fewer single quotes?